gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+      throws IOException {
+
+    int startRow = 0;
+    int count = 4;
+    final ExecutorService es = Executors.newFixedThreadPool(count);
+    int r = 0;
+    while (r < 50) {
+      r++;
+      List<Future<Integer>> results = new ArrayList<>(count);
+
+      for (int i = 0; i < count; ++i) {
+        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
+        final int start = startRow;
+        Future<Integer> submit = es.submit(new Callable<Integer>() {
+          @Override
+          public Integer call() {
+            LOG.info(String.format("Getting page size from %s...", start));
+            while (true) {
+              try {
+                List<String> pages =
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  //Figured out the size
+                  return pages.size() + start;
+                } else {
+                  //The size is not determined.
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
+              } catch (InterruptedException e) {
+                LOG.error(e.getMessage());
+              }
+            }
+          }
+        });
+        results.add(submit);
+        try {
+          Thread.sleep(250);
+        } catch (InterruptedException e) {
+          LOG.error(e.getMessage());
+        }
+      }
+
+      for (Future<Integer> result : results) {
+        try {
+          Integer integer = result.get(2, TimeUnit.MINUTES);
+          if (integer > 0) {
+            es.shutdownNow();
+            return integer;
+          }
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+          throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    throw new RuntimeException("Something seems wrong here. Having more than 5000*4*50 pages?");
+  }
+
+  /**
+   * Get all pages in an async mode.
+   */
+  private Collection<String> getPages(String startDate, String endDate, List<Dimension> dimensions,
+      ApiDimensionFilter countryFilter, Queue<Pair<String, FilterOperator>> toProcess) throws IOException {
+    String country = GoogleWebmasterFilter.countryFilterToString(countryFilter);
+
+    ConcurrentLinkedDeque<String> allPages = new ConcurrentLinkedDeque<>();
+    Random random = new Random();
+    //we need to retry many times because the path may be very long
+    final int retry = 120;
+    int r = 0;
+    while (r <= retry) {
```

Comment:
I am still not 100% sure what the code optimizes for. Is it number of concurrent requests, request rate/thread, total request rate, etc? I'll try to read your presentation more carefully but adding some comments in the code will help.

If you want to limit the number of concurrent requests, a fixed thread pool will do.
If you want to control the request submission rate, I'd claim that single job queue is a better approach. 

The rate of job submission in the thread-pool need not be == the request submission rate. You can use a gobblin.util.limiter.RateBasedLimiter before sending the request. That will make sure that you don't submit requests at a higher rate. That will guarantee you that you don't go over the limit yet you'll have more consistent flow of requests. I find the current Sleep-based approach fragile and the submission rate can vary greatly depending on the latency to send a request.

In the current, round-based approach, a single round can be held up by a single request that is taking longer to process on the Google side. During that time, you are not sending any requests.